src/nfa/nfa-concatenation.cc

```diff
@@ -26,131 +26,50 @@ Nfa concatenate(const Nfa& lhs, const Nfa& rhs, bool use_epsilon,
     // Compute concatenation of given automata.
     // Concatenation will proceed in the order of the passed automata: Result is 'lhs . rhs'.
 
-    if (lhs.initial.empty() || lhs.final.empty() || rhs.initial.empty()) { return Nfa{}; }
+    if (lhs.delta.post_size() == 0 || rhs.delta.post_size() == 0 || lhs.initial.empty() || lhs.final.empty() ||
+        rhs.initial.empty() || rhs.final.empty()) {
+        return Nfa{};
+    }
 
     const unsigned long lhs_states_num{lhs.delta.post_size() };
     const unsigned long rhs_states_num{rhs.delta.post_size() };
     Nfa result{}; // Concatenated automaton.
     StateToStateMap lhs_result_states_map_internal{}; // Map mapping rhs states to result states.
     StateToStateMap rhs_result_states_map_internal{}; // Map mapping rhs states to result states.
+    const bool lhs_accepts_empty_string{ is_in_lang(lhs, Run{{}, {}}) };
```

Comment:
Why not just write `lhs_accepts_empty_string = is_in_lang(...)`?

Reply:
Copy initialization (`int var = ...`) is a C concept which is discouraged in modern C++. Uniform initialization has added benefits such as consistent structure across variables of most types or prohibiting narrowing conversions. For reference, see [further explanation](https://www.learncpp.com/cpp-tutorial/variable-assignment-and-initialization/).

Otherwise, it is the same.